### ISSUE TITLE
trying to understand the impact of "super nodes"

### ISSUE THREAD

**Issue by brandenburg** [gnelissen/np-schedulability-analysis#25]
I was trying to see the impact of the new "super nodes" first-hand by running the version with the new bitfields on benchmarks you shared with and without super nodes. But I didn't get the expected results:

```
Benchmark 1: SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-log-uniform-discrete_69.csv
  Time (mean ± σ):     139.0 ms ±   0.7 ms    [User: 123.8 ms, System: 14.0 ms]
  Range (min … max):   138.1 ms … 140.5 ms    21 runs
 
Benchmark 2: SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-log-uniform-discrete_69.csv
  Time (mean ± σ):     139.2 ms ±   0.7 ms    [User: 123.8 ms, System: 14.2 ms]
  Range (min … max):   137.8 ms … 140.9 ms    21 runs
 
Summary
  SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-log-uniform-discrete_69.csv ran
    1.00 ± 0.01 times faster than SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-log-uniform-discrete_69.csv
```

```
Benchmark 1: SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-log-uniform-discrete_73.csv
  Time (mean ± σ):       8.2 ms ±   0.3 ms    [User: 6.6 ms, System: 1.2 ms]
  Range (min … max):     7.8 ms …   9.0 ms    295 runs
 
Benchmark 2: SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-log-uniform-discrete_73.csv
  Time (mean ± σ):       8.2 ms ±   0.3 ms    [User: 6.6 ms, System: 1.2 ms]
  Range (min … max):     7.7 ms …   9.1 ms    292 runs
 
Summary
  SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-log-uniform-discrete_73.csv ran
    1.00 ± 0.04 times faster than SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-log-uniform-discrete_73.csv
```

```
Benchmark 1: SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-log-uniform-discrete_94.csv
  Time (mean ± σ):       5.2 ms ±   0.2 ms    [User: 4.0 ms, System: 0.9 ms]
  Range (min … max):     4.8 ms …   6.3 ms    421 runs
 
  Warning: Command took less than 5 ms to complete. Note that the results might be inaccurate because hyperfine can not calibrate the shell startup time much more precise than this limit. You can try to use the `-N`/`--shell=none` option to disable the shell completely.
 
Benchmark 2: SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-log-uniform-discrete_94.csv
  Time (mean ± σ):       5.2 ms ±   0.2 ms    [User: 4.0 ms, System: 0.9 ms]
  Range (min … max):     4.9 ms …   6.2 ms    391 runs
 
  Warning: Command took less than 5 ms to complete. Note that the results might be inaccurate because hyperfine can not calibrate the shell startup time much more precise than this limit. You can try to use the `-N`/`--shell=none` option to disable the shell completely.
 
Summary
  SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-log-uniform-discrete_94.csv ran
    1.00 ± 0.06 times faster than SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-log-uniform-discrete_94.csv
```

```
Benchmark 1: SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-uniform-discrete_29.csv
  Time (mean ± σ):      99.6 ms ±   1.2 ms    [User: 88.6 ms, System: 10.0 ms]
  Range (min … max):    98.1 ms … 104.1 ms    29 runs
 
Benchmark 2: SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-uniform-discrete_29.csv
  Time (mean ± σ):      99.0 ms ±   1.1 ms    [User: 88.3 ms, System: 9.8 ms]
  Range (min … max):    97.8 ms … 102.3 ms    29 runs
 
Summary
  SAG-bitfield/build/nptest --sn=0 benchmarks/jobset-uniform-discrete_29.csv ran
    1.01 ± 0.02 times faster than SAG-bitfield/build/nptest --sn=1 benchmarks/jobset-uniform-discrete_29.csv
```

The "super nodes" don't seem to have any effect... which seems wrong. Am I doing something wrong or are the benchmarks too small to reveal a positive effect?

**Comment by porya-gohary** [comment]
I am not the one who invented and implemented the supernode.
@gnelissen knows better.
But if I am not mistaken to see the difference you have to compare it with the master branch (I am not sure about `--sn` option).

**Comment by brandenburg** [comment]
Ok, thanks. Looking forward to what Geoffrey has to say. 

In the meantime, I compared `master` and `v3.0`, but I still must be doing something wrong: the version in `master` seems to faster!?

```
Benchmark 1: SAG-v3/build/nptest benchmarks/jobset-log-uniform-discrete_69.csv
  Time (mean ± σ):     139.2 ms ±   0.5 ms    [User: 123.9 ms, System: 14.1 ms]
  Range (min … max):   138.3 ms … 140.1 ms    21 runs
 
Benchmark 2: SAG-master/build/nptest  benchmarks/jobset-log-uniform-discrete_69.csv
  Time (mean ± σ):     104.4 ms ±   0.3 ms    [User: 97.8 ms, System: 5.9 ms]
  Range (min … max):   103.6 ms … 105.1 ms    28 runs
 
Summary
  SAG-master/build/nptest  benchmarks/jobset-log-uniform-discrete_69.csv ran
    1.33 ± 0.01 times faster than SAG-v3/build/nptest benchmarks/jobset-log-uniform-discrete_69.csv
```

```
Benchmark 1: SAG-v3/build/nptest benchmarks/jobset-log-uniform-discrete_73.csv
  Time (mean ± σ):       8.2 ms ±   0.3 ms    [User: 6.6 ms, System: 1.2 ms]
  Range (min … max):     7.8 ms …   9.3 ms    290 runs
 
Benchmark 2: SAG-master/build/nptest  benchmarks/jobset-log-uniform-discrete_73.csv
  Time (mean ± σ):       6.6 ms ±   0.2 ms    [User: 5.4 ms, System: 0.9 ms]
  Range (min … max):     6.3 ms …   7.7 ms    354 runs
 
Summary
  SAG-master/build/nptest  benchmarks/jobset-log-uniform-discrete_73.csv ran
    1.24 ± 0.06 times faster than SAG-v3/build/nptest benchmarks/jobset-log-uniform-discrete_73.csv
```

```
Benchmark 1: SAG-v3/build/nptest benchmarks/jobset-log-uniform-discrete_94.csv
  Time (mean ± σ):       5.2 ms ±   0.2 ms    [User: 4.0 ms, System: 0.9 ms]
  Range (min … max):     4.9 ms …   6.1 ms    386 runs
 
  Warning: Command took less than 5 ms to complete. Note that the results might be inaccurate because hyperfine can not calibrate the shell startup time much more precise than this limit. You can try to use the `-N`/`--shell=none` option to disable the shell completely.
 
Benchmark 2: SAG-master/build/nptest  benchmarks/jobset-log-uniform-discrete_94.csv
  Time (mean ± σ):       4.3 ms ±   0.2 ms    [User: 3.3 ms, System: 0.7 ms]
  Range (min … max):     3.9 ms …   5.3 ms    487 runs
 
  Warning: Command took less than 5 ms to complete. Note that the results might be inaccurate because hyperfine can not calibrate the shell startup time much more precise than this limit. You can try to use the `-N`/`--shell=none` option to disable the shell completely.
 
Summary
  SAG-master/build/nptest  benchmarks/jobset-log-uniform-discrete_94.csv ran
    1.22 ± 0.08 times faster than SAG-v3/build/nptest benchmarks/jobset-log-uniform-discrete_94.csv
```

```
Benchmark 1: SAG-v3/build/nptest benchmarks/jobset-uniform-discrete_29.csv
  Time (mean ± σ):      99.6 ms ±   0.7 ms    [User: 88.4 ms, System: 10.2 ms]
  Range (min … max):    98.2 ms … 100.8 ms    29 runs
 
Benchmark 2: SAG-master/build/nptest  benchmarks/jobset-uniform-discrete_29.csv
  Time (mean ± σ):      73.1 ms ±   0.4 ms    [User: 68.1 ms, System: 4.4 ms]
  Range (min … max):    72.5 ms …  74.1 ms    39 runs
 
Summary
  SAG-master/build/nptest  benchmarks/jobset-uniform-discrete_29.csv ran
    1.36 ± 0.01 times faster than SAG-v3/build/nptest benchmarks/jobset-uniform-discrete_29.csv
```

**Comment by porya-gohary** [comment]
@brandenburg I just noticed one thing. Do you run the test with  `-m 4` option? 🤔

**Comment by brandenburg** [comment]
I'm currently benchmarking the WATERS17 Core 2 workload. I don't have final numbers yet, but a clear difference is visible already.

I suppose so far I've just been measuring that it's really fast to find the first deadline miss on a massively overloaded system no matter what you do... :-/

**Comment by gnelissen** [comment]
> The "super nodes" don't seem to have any effect... which seems wrong. Am I doing something wrong or are the benchmarks too small to reveal a positive effect?

@brandenburg as mentioned by @porya-gohary, you will see the real difference by comparing the master branch and the v3.0. The v3.0 introduces much more than a new level of hierarchy in the graph. It also (1) cache information at those different levels, (2) re-organize the explore function to avoid re-computation.

The --sn option does not deactivate those optimizations. It only forces to have a single state per node. The difference with --sn activated or not will really show performance differences for multicore scheduling of DAGs, self-suspending tasks and even more for the analysis of TSN time aware shapers (which is not part of the v3.0 but is its own a separate project developed by @nidhi2396).

Now, specifically for the analysis of single core, supernodes have no effect at all because the original analysis already merged all states with the same set of dispatched jobs. That property was never formally proven, but it is empirically true. Therefore, there is anyway a single state per node. It is the caching and new organization of the explore function that causes the performance improvement for that case.

**Comment by brandenburg** [comment]
Here's a comparison of the `master` vs. `v3.0` branches on macOS  running on an Apple M1 chipx for the WATERS17 workload with and without spin delay as prepared and kindly shared by Mo Foughali. Based on 10 runs each:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `SAG-v3/build/nptest Mo/waters_core2.csv` | 119.145 ± 1.293 | 117.885 | 122.483 | 1.00 |
| `SAG-master/build/nptest  Mo/waters_core2.csv` | 572.815 ± 12.120 | 548.649 | 585.272 | 4.81 ± 0.11 |


| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `SAG-v3/build/nptest Mo/waters_core2_overhead.csv` | 119.453 ± 1.107 | 117.332 | 121.357 | 1.00 |
| `SAG-master/build/nptest  Mo/waters_core2_overhead.csv` | 570.320 ± 15.778 | 550.184 | 587.845 | 4.77 ± 0.14 |

This shows: 
1. A significant speedup of around 4.8 for this workload (but not quite "10x"). 
2. There is no significant difference between the workload with and without spin-delay inflation. 

As this was done on macOS, I don't expect the picture to change when the bitfield patch is merged.

**Comment by gnelissen** [comment]
I recognize these numbers for the waters case study. Those are the numbers you obtain when the clever trick Foughali et al. uses in their UPPAAL model is not applied for the SAG, namely that the highest priority task cannot be preempted. Since it cannot be preempted, there should not be any preemption point between its segments. Foughali et al. enforces this in their UPPAAL model, so it is fair to apply it in our inputs to the SAG too. The runtime will then decrease significantly. 
I did not have time to process the input files to do that yet.

In the future we could make that check directly in the SAG analysis instead of expecting the user to provide the appropriate inputs, but that's future work. One more item on the todo list.

**Comment by brandenburg** [comment]
Interesting. I don't think I fully get it yet. Do you have a small example you could share? It would be interesting to look at a collected graph. 

I don't really understand why the highest-priority job's segments, even if there are many, would cause superfluous interleavings. On a uniprocessor, there's only one eligible next segment, right? How would such a state have multiple successors?  In other words, I would have naively assumed that the existing exploration logic is efficient in suppressing impossible interleavings for the highest-priority job.

**Comment by brandenburg** [comment]
To satisfy my own curiosity, I ran (on Linux) `master` with the bitfields patch vs . `v3` with the bitfields patch. The realized speedup is heavily dependent on the difficulty of the problem — the longer the original runtime, the better the speedup, it seems.  So there's no simple "rule of thumb" way of relating the `v3` version with the older version, with speedups all over the board ranging from <10x to >100x, but the gains are truly amazing.  

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `SAG-bitfield/build/nptest -m 4 benchmarks/jobset-log-uniform-discrete_69.csv` | 3.775 ± 0.015 | 3.749 | 3.795 | 1.00 |
| `SAG-master+bitfield/build/nptest -m 4 benchmarks/jobset-log-uniform-discrete_69.csv` | 933.374 ± 28.584 | 887.340 | 961.860 | 247.24 ± 7.64 |

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `SAG-bitfield/build/nptest -m 4 benchmarks/jobset-log-uniform-discrete_73.csv` | 330.3 ± 5.7 | 315.7 | 336.5 | 1.00 |
| `SAG-master+bitfield/build/nptest -m 4 benchmarks/jobset-log-uniform-discrete_73.csv` | 3409.8 ± 56.2 | 3344.4 | 3510.0 | 10.32 ± 0.25 |

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `SAG-bitfield/build/nptest -m 4 benchmarks/jobset-log-uniform-discrete_94.csv` | 235.8 ± 7.1 | 228.4 | 246.3 | 1.00 |
| `SAG-master+bitfield/build/nptest -m 4 benchmarks/jobset-log-uniform-discrete_94.csv` | 1719.3 ± 10.4 | 1696.3 | 1731.6 |  7.29 ± 0.22 |

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `SAG-bitfield/build/nptest -m 4 benchmarks/jobset-uniform-discrete_29.csv` | 495.1 ± 10.0 | 476.7 | 505.6 | 1.00 |
| `SAG-master+bitfield/build/nptest -m 4 benchmarks/jobset-uniform-discrete_29.csv` | 68855.8 ± 1248.9 | 65461.3 | 69680.2 | 139.07 ± 3.77|

Just look at the first one: a > 200x improvement! How is that even possible? Now I really, really want to read the paper on it. :-)

**Comment by gnelissen** [comment]
> Interesting. I don't think I fully get it yet. Do you have a small example you could share? It would be interesting to look at a collected graph.
> 
> I don't really understand why the highest-priority job's segments, even if there are many, would cause superfluous interleavings. On a uniprocessor, there's only one eligible next segment, right? How would such a state have multiple successors? In other words, I would have naively assumed that the existing exploration logic is efficient in suppressing impossible interleavings for the highest-priority job.

The problem is not with the number of nodes or states generated. As you stated, there is only one outgoing edge after each state resulting from a the dispatch of a highest-priority segment (except for the last segment of the highest priority task). The problem is that, at each preemption point, the analysis will still check all jobs that are ready in the same interval and decide whether they can be dispatched or not, eventually deciding for all of them except one that they cannot.

When I looked at the model of Foughali et al. I realized they have a different model for the highest-priority task. They do not call the scheduler model after each of its segment is dispatched, therefore they do not do any check on which other segments are ready and may be dispatched. That saves a lot of computation time that way. 
We can do the same with the SAG by modeling the highest priority task as a non-preemptive one. You will have the same output in a fraction of the time.

**Comment by Mo-F** [comment]
I realised a quick run, and second Björn's results. On my Macbook Air M1, WATERS Core 2 with blocking:

master: 750.718070s
v3: 157.902713s

As for further speedup based on my model, I'm happy we can inspire each other. 

@Geoffrey: I have not forgotten about your emails. I have a 9-hour flight on Sunday, I will use the time to reply and also send the required files.

**Comment by gnelissen** [comment]
[waters_core2-overheads_preprocessed.csv](https://github.com/user-attachments/files/16913371/waters_core2-overheads_preprocessed.csv)
[waters_core2-overheads-prec_preprocessed.csv](https://github.com/user-attachments/files/16913372/waters_core2-overheads-prec_preprocessed.csv)

@brandenburg @Mo-F I finally found a bit of time to update the input file such that it accounts for the fact that the highest priority task is not preemptable (same as the UPPAAL model of Mo). I attached them here for your reference. 
If you run the SAG analysis on those files you'll obtain the same BCRT and WCRT for the tasks than with the original ones, but the runtime is a fraction of what you got in your earlier tests.
Today, my runtime is 4.392s on my laptop.

**Comment by Mo-F** [comment]
@gnelissen bravo :)

Something is worth clarifying though. In my model, preventing the generation of preemption locations for the highest priority task is more aesthetic than anything else: I did not do it for performance concerns, for the simple reason that model-checking-wise, it has more or less no effect on verification time. That is, if you generate the preemption locations for the highest priority task anyways, the preemption channel will be always deactivated in the TA of such task, resulting in no effect on state-space computation.  

But as I said, I'm happy it inspired you.

**Comment by Mo-F** [comment]
> [waters_core2-overheads_preprocessed.csv](https://github.com/user-attachments/files/16913371/waters_core2-overheads_preprocessed.csv) [waters_core2-overheads-prec_preprocessed.csv](https://github.com/user-attachments/files/16913372/waters_core2-overheads-prec_preprocessed.csv)
> 
>  (same as the UPPAAL model of Mo). 

@gnelissen I checked your attached files, and the above quote is not entirely accurate. 

As I suspected, you are declaring the task with the highest priority as a single segment with BCET and WCET equaling, respectively, the sum of BCETs and WCETs with all segments in the task. I do not do have this "optimization" in my model for the simple reason that I do not want to lose information to verify other properties than schedulability or compute other bounds than WCRTs (for example a bound involving the execution of a particular segment in this very task). All I do is not generating preemption segments in this very task, which is aesthetic as I said. 

Still, I will prepare a model with this for you, and send it within the promised emails. However, I do not expect any major gain in verification time on my side.
